sql/catalyst/src/main/scala/org/apache/spark/sql/connector/catalog/CatalogV2Util.scala

```diff
@@ -431,4 +436,45 @@ private[sql] object CatalogV2Util {
       .getOrElse(catalogManager.v2SessionCatalog)
       .asTableCatalog
   }
+
+  def v2ColumnsToStructType(columns: Array[Column]): StructType = {
+    StructType(columns.map(v2ColumnToStructField))
+  }
+
+  def v2ColumnToStructField(col: Column): StructField = {
+    val metadata = Option(col.metadataInJSON()).map(Metadata.fromJson).getOrElse(Metadata.empty)
+    var f = StructField(col.name(), col.dataType(), col.nullable(), metadata)
+    Option(col.comment()).foreach { comment =>
+      f = f.withComment(comment)
+    }
+    Option(col.defaultValue()).foreach { default =>
+      val defaultSQL = CatalogV2Util.getDefaultValueSQL(default)
+      f = f.withExistenceDefaultValue(defaultSQL).withCurrentDefaultValue(defaultSQL)
+    }
+    f
+  }
+
+  def structTypeToV2Columns(schema: StructType): Array[Column] = {
+    schema.fields.map { f =>
+      // TODO: remove it when our testing v2 tables no longer implements and replies on the
+      //       deprecated `Table.schema`.
+      val defaultValue = f.getCurrentDefaultValue().map { sql =>
+        val e = CatalystSqlParser.parseExpression(sql)
+        if (!e.resolved || !e.foldable) {
+          // This should not happen as we do validation before setting column default values, but
+          // in case something goes wrong, we treat it as no default value.
+          null
```

Comment:
we should throw an exception here instead of returning NULL, since that would comprise an incorrect result.